include/storage/shared_monitor.hpp

```diff
@@ -0,0 +1,175 @@
+#include <boost/interprocess/sync/interprocess_condition.hpp>
+#include <boost/interprocess/sync/interprocess_mutex.hpp>
+#include <boost/interprocess/sync/interprocess_semaphore.hpp>
+#include <boost/interprocess/sync/scoped_lock.hpp>
+
+#if defined(__linux__)
+#define USE_BOOST_INTERPROCESS_CONDITION 1
+#endif
+
+namespace osrm
+{
+namespace storage
+{
+
+namespace
+{
+namespace bi = boost::interprocess;
+
+template <class Lock> class InvertedLock
+{
+    Lock &lock;
+
+  public:
+    InvertedLock(Lock &lock) : lock(lock) { lock.unlock(); }
+    ~InvertedLock() { lock.lock(); }
+};
+}
+
+// The shared monitor implementation based on a semaphore and mutex
+template <typename Data> struct SharedMonitor
+{
+    using mutex_type = bi::interprocess_mutex;
+
+    SharedMonitor(const Data &initial_data)
+    {
+        shmem = bi::shared_memory_object(bi::open_or_create, Data::name, bi::read_write);
+
+        bi::offset_t size = 0;
+        if (shmem.get_size(size) && size == 0)
+        {
+            shmem.truncate(internal_size + sizeof(Data));
+            region = bi::mapped_region(shmem, bi::read_write);
+            new (&internal()) InternalData;
+            new (&data()) Data(initial_data);
+        }
+        else
+        {
+            region = bi::mapped_region(shmem, bi::read_write);
+        }
+    }
+
+    SharedMonitor()
+    {
+        try
+        {
+            shmem = bi::shared_memory_object(bi::open_only, Data::name, bi::read_write);
+
+            bi::offset_t size = 0;
+            if (!shmem.get_size(size) || size != internal_size + sizeof(Data))
+            {
+                auto message =
+                    boost::format("Wrong shared memory block '%1%' size %2%, expected %3% bytes") %
+                    (const char *)Data::name % size % (internal_size + sizeof(Data));
+                throw util::exception(message.str() + SOURCE_REF);
+            }
+
+            region = bi::mapped_region(shmem, bi::read_write);
+        }
+        catch (const bi::interprocess_exception &exception)
+        {
+            auto message = boost::format("No shared memory block '%1%' found, have you forgotten "
+                                         "to run osrm-datastore?") %
+                           (const char *)Data::name;
+            throw util::exception(message.str() + SOURCE_REF);
+        }
+    }
+
+    Data &data() const
+    {
+        auto region_pointer = reinterpret_cast<char *>(region.get_address());
+        return *reinterpret_cast<Data *>(region_pointer + internal_size);
+    }
+
+    mutex_type &get_mutex() const { return internal().mutex; }
+
+#if USE_BOOST_INTERPROCESS_CONDITION
+    template <typename Lock> void wait(Lock &lock) { internal().condition.wait(lock); }
+
+    void notify_all() { internal().condition.notify_all(); }
+#else
+    template <typename Lock> void wait(Lock &lock)
+    {
+        if (((internal().head + 1) & (buffer_size - 1)) == (internal().tail & (buffer_size - 1)))
```

Comment:
What is the advantage of using `& (buffer_size - 1)` instead of `% buffer_size`? I'm more familiar with the latter when it comes to wrap-a-round indexing.

Wrapping the ring buffer in smaller helper functions to hide the index "magic" might improve readability.